### ISSUE TITLE
Feature Request: break down assemblies

### ISSUE THREAD

**Issue by jiatao99** [dotnetrdf/dotnetrdf#384]
Is it possible to follow .NET core design pattern to break down the big dotnetrdf assembly into smaller assemblies? Something like:

dotnetrdf.abstraction.dll:  interfaces
dotnetrdf.client.dll: client invoking library including QueryBuilder

Most user just need the two libraries above to connect external RDF database

dotnetrdf.parser.dll: SPARQL parser
dotnetrdf.server.dll: in memory RDF implementation

I know it is going to be a huge refactor, but I think it worth the effort to make it more efficient. 

What do you think?

**Comment by kal** [comment]
That was something I had thought about for the 3.0 release and it has been quite a long-standing issue (see https://github.com/dotnetrdf/dotnetrdf/issues/32 - the 1.9 that is referred to in there is work that Rob did a long time ago on a refactor but that work didn't end up being our 2.0 release). I think it is worth revisiting and thinking about the split that we would take. I think there are roughly these elements:

- Core Abstractions
- Parsing & Serialization
- In-memory store and in-memory SPARQL engine
- Standard protocol clients (SPARQL protocol mainly)
- Third party triple store integrations (a separate assembly for each integration?)
- Dynamic Graphs
- SPARQL Query Builder
- SPIN library
- Full-text indexing

Some of these might need to go together or make sense to go together so it might not end up being that many separate assemblies (for example your suggestion that SPARQL query builder should be included in a client library that would  also cover the remote protocols)

One thing to guide how we split things up would be to decided exactly why we are doing that. One reason is to make the size of install smaller. Another is to reduce the number of dependency libraries that are called in. So if isolating dependencies is an issue we might actually divide things up differently (for example the support for RDFa pulls in Anglesharp / HTMLAgilityPack so we could isolate the RDFa parser/serialiser into its own package so that dependency is also isolated).

What would your motivation be for doing this break down?

**Comment by kal** [comment]
Just for reference, this is the document that Rob created when planning out the 1.9 design with his original plan of the breakdown of modules (it doesn't cover everything that is in the library now though): https://github.com/dotnetrdf/dotnetrdf/wiki/DeveloperGuide-1.9-Libraries

**Comment by jiatao99** [comment]
Thanks. That's a very good document. A couple of suggestion:

1. dotNetRDF.Sparql.Core: only the basic logic to handle XML, JSON parsing and http connection. This is the minimum requirement to call SPARQL http store using string. Since SPARQL is SQL like language, maybe we can borrow ADO.Net Command, Connection, IDataReader syntax?
2. dotNetRDF.Sparql.Query: strong typed object model to wrap SPARQL
3. dotNetRDF.Sparql.Parser: parsing SPARQL into strong typed Query object

**Comment by kal** [comment]
That's a neat breakdown. I think maybe (1) in you list could be handled by a meta package that pulls in the core abstractions, protocol support and XML and JSON parsing. 

The idea of building on ADO.Net is interesting. The existing DataTables library provides some of that interop with ADO.NET, though only in terms of being able to read RDF into a data table and being able to treat a SPARQL Result Set as a data table. If there were some volunteers to work on making the integration more complete then that would be cool.

It is actually quite a good point also that some of the parsers introduce their own dependencies so are maybe good candidates for separating out, For example the RDFa parser requires additional libraries and I'm willing to bet that only a few people need an RDFa parser - so perhaps the Parsing & Serialization in my list should be broken down a bit too.

**Comment by kal** [comment]
# Goals of the restructuring

These are not in any particular order

 * Reduce the surface area of APIs for some core use-cases
    * Using the library to interact with standard remote services (Graph Store Protocol, SPARQL HTTP protocol, maybe later LDP?)
    * Simple read/process/write loops that don't use SPARQL
    
 * Isolate third-party dependencies
   * HTML parser dependency is only needed for RDFa IO
   * JSON parser dependency is only needed for RDF/JSON and JSON-LD processing
   * Full-text and Virtuoso dependencies are already isolated
   
 * Try to keep the organisation understandable
   * Avoid a big long list of NuGet packages that make it hard to figure out what pieces are needed
   * Meta-packages could help with this, and there could be a single dotNetRdf meta-package that combines them all as well as more focussed ones
   
# Current Situation

These are the current elements of dotNetRDF:

- Core RDF data model: triples, graphs
- Triplestore data model: (To be simplified compared to DNR2)
- SPARQL service interface - query interface

- RDF configuration model
  - The current implementation involves a base interface that is implemented across a lot of classes
  - might be hard to disentangle from core
  - factory classes in other will need to be registered from other assemblies

- Dynamic graph/nodes

- JSON-LD processing (JSON-LD algorithms)

- IO interfaces
    - there is also a central register of implementations (to support dispatching on MIME-type and/or file extension) so if implementations get separated out into other assemblies there will need to be some sort of registration process. This may end up being 

- IO implementations
    - GZip wrapper
    - Data URI wrapper
    - Loaders: FileLoader, EmbeddedResourceLoader
    - JSON-LD [depends on Newtonsoft.JSON]
    - N3 
    - NQuads
    - NTriples
    - RDFa [depends on HtmlAgilityPack/AngleSharp]
    - RDFJSON
    - RDF/XML
    - TriG
    - TriX
    - Turtle
    
- SPARQL query model
    - Algebra
    - Parser
    - Query Builder
    - Query optimiser
    - Extension functions (these might be too Leviathan-specific)
    
- SPARQL update model
    - Commands
    
- Leviathan Engine
    - Currently much of the engine logic is embedded in the algebra
    - SPARQL protocol processor
    
- Medusa Engine
    - new streaming engine, not yet implemented but likely 
    - would also need processor implementations unless the Leviathan ones can be suitably abstracted
    
- SHACL Validator

- SKOS API

- OWL API

- Store Implementations
    - Allegrograph
    - FourStore
    - Fuseki
    - Sesame
    - Stardog
    - SparqlProtocol
    - In-memory
    
- Web
    - old HTTP Context extensions
    Needs updating
    Web server implementation: 
        IHttpHandlers for SPARQL, Graph store protocol etc.
    
- Datatables Interop
    - dump triples as data table
    - access SPARQL results as a data table
    
- Virtuoso store implementation
    - separate due to additional dependency on Virtuoso client assembly

- Leviathan Full-Text extensions
    - separate due to additional dependency on Lucene.Net
    
- SPIN engine

# Proposed Structure

This is a proposed structure. The names are assembly / package names (Note change in capitalisation. In line with MS naming conventions, RDF should be capitalized as Rdf)

- dotNetRdf
  - A meta-package pulling in all the functionality that is currently in dotNetRDF 2.x's main package (see notes at the end)

- dotNetRdf.Core
    - Nodes, Triples, Graphs (interfaces and default in-memory implementations)
    - Core helpers
    - Triplestore interfaces
    - SPARQL interfaces
    - RDF Configuration
    - Core exceptions
    - Core I/O (all text and XML formats that can be supported with the .NET Standard 2.0 lib)
    
- dotNetRDF.Json
  - I/O for RDF/JSON and JSON-LD
  - JSON-LD Processor
  - DEPENDS ON: dotNetRdf.Core
  
- dotNetRdf.Client
    - HTTP-based store integrations (basically all of them except Virtuoso at this point)
    - Remote SPARQL client
    - DEPENDS ON: dotNetRdf.Core (though need to also install and configure dotNetRdf.JSON for JSON format support)

- dotNetRdf.Sparql
    - SPARQL Algebra
    - SPARQL Parser
    - SPARQL Query Builder
    - SPARQL Update Model
    - DEPENDS ON: dotNetRdf.Core
    
- dotNetRdf.InMemory
    - In-memory store implementation
    - Leviathan Engine + processors
    - (Later) Medusa Engine + processors
 
- dotNetRdf.Datatables
    - Datatables Interop

- dotNetRdf.Dynamic
    - Dynamic graph APIs

- dotNetRdf.Skos
    - Skos APIs

- dotNetRdf.Shacl
    - SHACL validator
    
- dotNetRdf.Owl
    - OWL APIs
    
- dotNetRdf.Spin
    - SPIN Inference engine

- dotNetRdf.AspNet
    - Legacy web stuff
    
- dotNetRdf.AspNetCore
    - (Later) New web stuff
    
- dotNetRdf.Virtuoso
    - Virtuoso store integration

Some sample scenarios under the new package structure:

- To read/process/write NTriples, Turtle, RDF/XML etc. files I need only `dotNetRdf.Core`
- If I want to handle JSON-LD I need `dotNetRdf.Json` (which will pull in Core)
- If I want to interact with a remote server (including sending SPARQL to remote servers) I need `dotNetRdf.Client` (which will pull in Core), and optionally dotNetRdf.Json if I want to handle JSON formats.
    - BUT if I also want to pre-parse and validate SPARQL queries I will need to also pull in `dotNetRdf.Sparql`
- If I want to interact with a remote Virtuoso server, I need `dotNetRdf.Virtuoso` (which will pull in dotNetRdf.Client and dotNetRdf.Core)
- If I want to query RDF locally I need `dotNetRdf.InMemory` (plus `dotNetRdf.Json` if I want JSON I/O)
- SHACL, OWL, SKOS, SPIN all require their own separate packages to be pulled in
- Full-Text query in the in-memory SPARQL engine requires its own package
- ASP.NET support requires its own package
- When we get around to adding ASP.NET Core support that will also be its own package

In addition to the above packages we can also define a `dotNetRdf` meta-package that would pull in (almost) everything that is currently covered by dotNetRDF  2.x (Core, Json, Client, Sparql, InMemory, Dynamic, Skos, Shacl, Owl). This would leave out a bit of the ASP.NET support but I don't think that's a thing that is widely used in any case.

**Comment by kal** [comment]
Just to note that after working with the above structure for a while on the release/3.0.0 branch, I found that the separation of "core" and "in memory" packages was really quite painful as most of the time I want SPARQL support locally as well as / instead of using a remote SPARQL service. So I've merged the in-memory store and SPARQL engine into the core library. This means that I can now read, write and query RDF files with just the dotNetRdf.Core assembly.
